File: rewrite-history.py
#!/usr/bin/env python3
"""Rewrite git history: change email, remove copilot mentions, shift to weekends."""

import subprocess
import re
import math
from datetime import datetime, timedelta

def run(cmd):
    return subprocess.run(cmd, capture_output=True, text=True, check=True)

# Get all commits on main in chronological order
result = run(['git', '--no-pager', 'log', 'main', '--reverse', '--format=%H|||%aI|||%s'])
commits = []
for line in result.stdout.strip().split('\n'):
    parts = line.split('|||')
    if len(parts) == 3:
        commits.append({'hash': parts[0], 'date': parts[1], 'msg': parts[2]})

print(f'Total commits on main: {len(commits)}')

# Generate weekend days (Sat+Sun) to spread commits across
weekends = []
start = datetime(2026, 3, 28)  # a Saturday
end = datetime(2026, 5, 17)
d = start
while d <= end:
    if d.weekday() == 5:  # Saturday
        weekends.append(d)
        weekends.append(d + timedelta(days=1))  # Sunday
    d += timedelta(days=1)

print(f'Available weekend days: {len(weekends)}')

# Distribute commits across weekend days
n = len(commits)
n_days = len(weekends)
# Spread evenly, multiple commits per day if needed
commits_per_day = max(1, math.ceil(n / n_days))

# Build hash -> (new_date, new_msg) mapping
NEW_EMAIL = 'tiongl@users.noreply.github.com'
NEW_NAME = 'Tiong Lee'

date_map = {}
msg_map = {}

day_idx = 0
count_on_day = 0

for i, c in enumerate(commits):
    # Assign weekend date with time spread across the day
    weekend_day = weekends[day_idx]
    # Spread commits within the day (9am - 11pm)
    hour_start = 9
    hour_end = 23
    if commits_per_day > 1:
        frac = count_on_day / commits_per_day
        hour = hour_start + frac * (hour_end - hour_start)
    else:
        hour = 10 + (i % 12)
    
    h = int(hour)
    m = int((hour - h) * 60)
    s = (i * 17) % 60  # pseudo-random seconds
    
    new_dt = weekend_day.replace(hour=h, minute=m, second=s)
    date_str = new_dt.strftime('%Y-%m-%dT%H:%M:%S') + '-07:00'
    date_map[c['hash']] = date_str
    
    # Fix copilot mentions in message
    msg = c['msg']
    msg = re.sub(r'[Cc]opilot\s+CLI\s+', 'CLI ', msg)
    msg = re.sub(r'[Cc]opilot\s+', '', msg)
    msg = re.sub(r'copilot', '', msg, flags=re.IGNORECASE)
    # Clean up any double spaces
    msg = re.sub(r'  +', ' ', msg).strip()
    if msg != c['msg']:
        print(f'  MSG: "{c["msg"]}" -> "{msg}"')
    msg_map[c['hash']] = msg
    
    count_on_day += 1
    if count_on_day >= commits_per_day:
        count_on_day = 0
        day_idx += 1
        if day_idx >= n_days:
            day_idx = n_days - 1

# Write env-filter script
env_lines = []
for h, d in date_map.items():
    env_lines.append(f'if [ "$GIT_COMMIT" = "{h}" ]; then')
    env_lines.append(f'  export GIT_AUTHOR_DATE="{d}"')
    env_lines.append(f'  export GIT_COMMITTER_DATE="{d}"')
    env_lines.append(f'fi')

env_script = '\n'.join(env_lines)
env_script += f'\nexport GIT_AUTHOR_EMAIL="{NEW_EMAIL}"'
env_script += f'\nexport GIT_COMMITTER_EMAIL="{NEW_EMAIL}"'
env_script += f'\nexport GIT_AUTHOR_NAME="{NEW_NAME}"'
env_script += f'\nexport GIT_COMMITTER_NAME="{NEW_NAME}"'

with open('env-filter.sh', 'w', newline='\n') as f:
    f.write(env_script)

# Write msg-filter script  
msg_lines = ['#!/bin/bash', 'read -r -d "" MSG || true', 'COMMIT="$GIT_COMMIT"']
for h, m in msg_map.items():
    if m != next((c['msg'] for c in commits if c['hash'] == h), ''):
        escaped = m.replace('"', '\\"').replace("'", "'\\''")
        msg_lines.append(f'if [ "$COMMIT" = "{h}" ]; then echo "{escaped}"; exit 0; fi')
msg_lines.append('echo "$MSG"')

with open('msg-filter.sh', 'w', newline='\n') as f:
    f.write('\n'.join(msg_lines))

print(f'\nGenerated env-filter.sh and msg-filter.sh')
print(f'Commits remapped to {day_idx + 1} weekend days')
print(f'\nDate range: {weekends[0].strftime("%Y-%m-%d")} to {weekends[day_idx].strftime("%Y-%m-%d")}')

# Show sample mappings
print('\nSample date mappings:')
for c in commits[:3] + commits[-3:]:
    old_d = c['date'][:10]
    new_d = date_map[c['hash']][:10]
    print(f'  {c["hash"][:8]} {old_d} -> {new_d}  {msg_map[c["hash"]][:60]}')

File: rewrite-history.js
const { execSync } = require('child_process');
const fs = require('fs');

const NEW_EMAIL = 'tiongl@users.noreply.github.com';
const NEW_NAME = 'Tiong Lee';

// Get all commits on main in chronological order
const log = execSync('git --no-pager log main --reverse --format=%H|||%aI|||%s', { encoding: 'utf8' });
const commits = log.trim().split('\n').map(line => {
  const [hash, date, ...msgParts] = line.split('|||');
  return { hash, date, msg: msgParts.join('|||') };
}).filter(c => c.hash);

console.log(`Total commits: ${commits.length}`);

// Generate weekend days (Sat+Sun) from Mar 28 to May 17, 2026
const weekends = [];
let d = new Date(2026, 2, 28); // Mar 28 (Sat)
const end = new Date(2026, 4, 17);
while (d <= end) {
  if (d.getDay() === 6) { // Saturday
    weekends.push(new Date(d));
    const sun = new Date(d);
    sun.setDate(sun.getDate() + 1);
    weekends.push(sun);
  }
  d.setDate(d.getDate() + 1);
}
console.log(`Weekend days available: ${weekends.length}`);

// Distribute commits across weekend days
const perDay = Math.ceil(commits.length / weekends.length);
let dayIdx = 0, countOnDay = 0;

const dateMap = {};
const msgMap = {};

for (let i = 0; i < commits.length; i++) {
  const c = commits[i];
  const wd = weekends[dayIdx];
  
  // Spread commits within day (9am-11pm)
  const hourFrac = 9 + (countOnDay / Math.max(perDay, 1)) * 14;
  const h = Math.floor(hourFrac);
  const m = Math.floor((hourFrac - h) * 60);
  const s = (i * 17) % 60;
  
  const pad = (n) => String(n).padStart(2, '0');
  const dateStr = `${wd.getFullYear()}-${pad(wd.getMonth()+1)}-${pad(wd.getDate())}T${pad(h)}:${pad(m)}:${pad(s)}-07:00`;
  dateMap[c.hash] = dateStr;
  
  // Fix copilot mentions
  let msg = c.msg;
  msg = msg.replace(/Copilot CLI /gi, 'CLI ');
  msg = msg.replace(/copilot\s+/gi, '');
  msg = msg.replace(/copilot/gi, '');
  msg = msg.replace(/  +/g, ' ').trim();
  
  if (msg !== c.msg) {
    console.log(`  MSG: "${c.msg}" -> "${msg}"`);
  }
  msgMap[c.hash] = msg;
  
  countOnDay++;
  if (countOnDay >= perDay) {
    countOnDay = 0;
    dayIdx++;
    if (dayIdx >= weekends.length) dayIdx = weekends.length - 1;
  }
}

// Build env-filter script for git filter-branch
let envScript = '';
for (const [hash, date] of Object.entries(dateMap)) {
  envScript += `if [ "$GIT_COMMIT" = "${hash}" ]; then\n`;
  envScript += `  export GIT_AUTHOR_DATE="${date}"\n`;
  envScript += `  export GIT_COMMITTER_DATE="${date}"\n`;
  envScript += `fi\n`;
}
envScript += `export GIT_AUTHOR_EMAIL="${NEW_EMAIL}"\n`;
envScript += `export GIT_COMMITTER_EMAIL="${NEW_EMAIL}"\n`;
envScript += `export GIT_AUTHOR_NAME="${NEW_NAME}"\n`;
envScript += `export GIT_COMMITTER_NAME="${NEW_NAME}"\n`;

fs.writeFileSync('env-filter.sh', envScript, { encoding: 'utf8' });

// Build msg-filter script
let msgScript = '#!/bin/bash\nread -r -d "" MSG || true\n';
for (const c of commits) {
  if (msgMap[c.hash] !== c.msg) {
    const escaped = msgMap[c.hash].replace(/"/g, '\\"');
    msgScript += `if [ "$GIT_COMMIT" = "${c.hash}" ]; then echo "${escaped}"; exit 0; fi\n`;
  }
}
msgScript += 'echo "$MSG"\n';

fs.writeFileSync('msg-filter.sh', msgScript, { encoding: 'utf8' });

console.log(`\nGenerated env-filter.sh and msg-filter.sh`);
console.log(`Commits spread across ${dayIdx + 1} weekend days`);
console.log(`\nSample mappings:`);
const samples = [...commits.slice(0, 3), ...commits.slice(-3)];
for (const c of samples) {
  console.log(`  ${c.hash.slice(0,8)} ${c.date.slice(0,10)} -> ${dateMap[c.hash].slice(0,10)}  ${msgMap[c.hash].slice(0,50)}`);
}
